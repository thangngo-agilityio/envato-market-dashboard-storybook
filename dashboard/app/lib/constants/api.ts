// TODO: Will provide latter
export const BASE_API = '';
export const USERS_API = process.env.NEXT_PUBLIC_API_USER || BASE_API;
export const STATISTICAL_API =
  process.env.NEXT_PUBLIC_STATISTICAL_API || BASE_API;

export const END_POINTS = {
  AUTHEN_USERS: '/authenUsers',
  USERS: '/users',
  TRANSACTIONS: '/transactions',
  PRODUCTS: '/products',
  RECENT_ACTIVITIES: '/activities',
  STATISTICS: '/statistics',
  REVENUE: '/revenue',
  EFFICIENCY: '/efficiency',
  NOTIFICATION: '/notifications',
  EMPLOYEES: '/employees',
  OVERALL_BALANCE: '/overall-balance',
  SIGN_IN: '/sign-in',
  SIGN_UP: '/sign-up',
  DELETE_TRANSACTION: '/delete-transactions',
  EDIT_TRANSACTION: '/edit-transactions',
  CREATE_PIN: '/create-pin-code',
  CONFIRM_PIN: '/pin-code',
  UPDATE_PASSWORD: '/update-password',
  MY_WALLET: '/wallets',
  ADD_MONEY: '/add-money',
  SEND_MONEY: '/send-money',
  SUPPORT: '/supports',
  LOCK: '/lock',
  UNLOCK: '/unlock',
  ADMIN: '/admin',
  SETTINGS: '/settings',
  LOGIN: '/login',
  WALLET: '/my-wallets',
};

export const SEARCH_PARAM = {
  EMAIL: 'email',
  PASSWORD: 'password',
  TRANSACTION_NAME: 'transactionName',
};
