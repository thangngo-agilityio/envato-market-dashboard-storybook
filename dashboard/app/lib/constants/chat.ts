export const FIREBASE_CHAT = {
  USERS: 'users',
  USER_CHATS: 'userChats',
  CHATS: 'chats',
};

export const USER_CHATS_FIELD = {
  USER_INFO: '.userInfo',
  DATE: '.date',
  LAST_MESSAGE: '.lastMessage',
};
