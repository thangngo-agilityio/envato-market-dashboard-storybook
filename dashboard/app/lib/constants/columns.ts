// Types
import {
  TDataSource,
  TProduct,
  TProductResponse,
  TRecentActivities,
  TTransaction,
} from '../interfaces';

export const COLUMNS_DASHBOARD = (
  onRenderHead: (title: string, key: string) => void,
  onRenderRole: (role: TTransaction) => void,
  onRenderBody: ({ id, image, name }: TDataSource) => void,
  onRenderActionIcon: (data: TTransaction) => void,
  onRenderSpent: (spent: TTransaction) => void,
  onRenderEmail: (email: TTransaction) => void,
) => [
  {
    title: 'Customer name',
    key: 'name',
    renderHead: onRenderHead,
    renderBody: onRenderBody,
  },
  {
    title: 'Email',
    key: 'email',
    renderHead: onRenderHead,
    renderBody: onRenderEmail,
  },
  {
    title: 'Location',
    key: 'location',
    renderHead: onRenderHead,
  },
  {
    title: 'Spent',
    key: 'spent',
    renderHead: onRenderHead,
    renderBody: onRenderSpent,
  },
  {
    title: 'Role',
    key: 'role',
    renderHead: onRenderHead,
    renderBody: onRenderRole,
  },
  {
    title: '',
    key: 'action',
    renderBody: onRenderActionIcon,
    renderHead: onRenderHead,
  },
];

export const COLUMNS_HISTORY = (
  renderHead: (title: string, key: string) => void,
  renderBody: ({ id, image, name }: TDataSource) => void,
  renderPaymentStatus: ({ paymentStatus }: TDataSource) => void,
  renderTransactionStatus: ({ paymentStatus }: TDataSource) => void,
  renderSpent: (spent: TTransaction) => void,
) => [
  {
    title: 'Customer name',
    key: 'name',
    renderHead: renderHead,
    renderBody: renderBody,
  },
  {
    title: 'Date',
    key: 'date',
    renderHead: renderHead,
  },
  {
    title: 'Amount',
    key: 'spent',
    renderHead: renderHead,
    renderBody: renderSpent,
  },
  {
    title: 'Payment',
    key: 'paymentStatus',
    renderHead: renderHead,
    renderBody: renderPaymentStatus,
  },
  {
    title: 'Status',
    key: 'transactionStatus',
    renderHead: renderHead,
    renderBody: renderTransactionStatus,
  },
];

export const COLUMNS_PRODUCTS = (
  onRenderHead: (title: string, key: string) => void,
  onRenderBody: ({ id, name }: TDataSource) => void,
  onRenderGallery: ({ imageURLs, name }: TProduct) => void,
  onRenderPrice: (amount: TProduct) => void,
  onRenderStatus: (productStatus: TDataSource) => void,
  onRenderQuantity: (stock: TProduct) => void,
  onRenderActionIcon: (data: TProductResponse) => void,
) => [
  {
    title: 'Product Name',
    key: 'name',
    renderHead: onRenderHead,
    renderBody: onRenderBody,
  },
  {
    title: 'Gallery Thumbnail',
    key: 'imageURLs',
    renderHead: onRenderHead,
    renderBody: onRenderGallery,
  },
  {
    title: 'Price',
    key: 'price',
    renderHead: onRenderHead,
    renderBody: onRenderPrice,
  },
  {
    title: 'Status',
    key: 'productStatus',
    renderHead: onRenderHead,
    renderBody: onRenderStatus,
  },
  {
    title: 'Quantity',
    key: 'quantity',
    renderHead: onRenderHead,
    renderBody: onRenderQuantity,
  },
  {
    title: 'Date',
    key: 'date',
    renderHead: onRenderHead,
  },
  {
    title: '',
    key: 'action',
    renderBody: onRenderActionIcon,
    renderHead: onRenderHead,
  },
];

export const COLUMNS_RECENT_ACTIVITIES = (
  onRenderHead: (title: string, key: string) => void,
  onRenderBody: ({ id, actionName }: TDataSource) => void,
  onRenderName: (actionName: TRecentActivities) => void,
  onRenderEmail: (email: TRecentActivities) => void,
) => [
  {
    title: 'ID Action',
    key: '_id',
    renderHead: onRenderHead,
    renderBody: onRenderBody,
  },
  {
    title: 'Action Name',
    key: 'actionName',
    renderHead: onRenderHead,
    renderBody: onRenderName,
  },
  {
    title: 'Email',
    key: 'email',
    renderHead: onRenderHead,
    renderBody: onRenderEmail,
  },
  {
    title: 'Date',
    key: 'date',
    renderHead: onRenderHead,
  },
];
