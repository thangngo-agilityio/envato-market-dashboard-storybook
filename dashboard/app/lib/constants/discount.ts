export const DEFAULT_DISCOUNT_PERCENTAGE = 0.6;
export const LIMIT_OF_BONUS = 5;
