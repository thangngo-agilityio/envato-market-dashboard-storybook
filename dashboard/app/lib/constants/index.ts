export * from './routers';
export * from './time';
export * from './images';
export * from './form';
export * from './messages';
export * from './regex';
export * from './texts';
export * from './themes';
export * from './api';
export * from './images';
export * from './menu';
export * from './sidebar';
export * from './charts';
export * from './sidebar';
export * from './common';
export * from './options';
export * from './transaction';
export * from './store';
export * from './status';
export * from './pagination';
export * from './columns';
export * from './faq';
export * from './avatar';
export * from './sizes';
export * from './setting';
export * from './termAndCondition';
export * from './notification';
export * from './authentication';
export * from './discount';
export * from './metadata';
export * from './og-properties';
export * from './chat';
export * from './product';
