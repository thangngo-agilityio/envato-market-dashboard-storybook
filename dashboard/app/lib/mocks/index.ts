export * from './spending';
export * from './table';
export * from './transaction';
export * from './product';
export * from './user';
export * from './time';
export * from './statistic';
export * from './chat';
export * from './issues';
export * from './wallet';
export * from './supports';
export * from './recentActivities';
