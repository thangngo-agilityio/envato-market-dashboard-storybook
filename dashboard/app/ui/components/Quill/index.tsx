import dynamic from 'next/dynamic';

import { useCallback, useEffect, useId } from 'react';
import { VStack, Flex, Text, useColorModeValue, theme } from '@chakra-ui/react';
import 'react-quill/dist/quill.snow.css';
import { Controller, useForm } from 'react-hook-form';
const ReactQuill = dynamic(() => import('react-quill'), { ssr: false });

// Components
import { SendIconLight } from '..';
import CustomButton from '@/ui/components/common/Button';

// Interfaces
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// bases
import { colors } from '@/ui/themes/bases';

// Hooks
import { sendMessage } from '@/lib/utils';

interface QuillProps {
  userUid: string | undefined;
  avatarUser: string;
  nameUser: string;
}

const Quill = ({ userUid, avatarUser, nameUser }: QuillProps) => {
  const currentUser = authStore((state) => state.user);

  // TODO: get from list users
  // const senderId = user?.id || '';

  // TODO: if have real id from database
  // const idRoomChat = useGetRoomChat();

  const {
    control,
    formState: {
      errors: { root },
      isValid,
      isSubmitting,
    },
    handleSubmit,
    reset,
  } = useForm<TMessages>({
    defaultValues: {
      text: '',
      id: useId(),
      senderId: currentUser?.uid,
      date: { nanoseconds: 0, seconds: 0 },
    },
    mode: 'onBlur',
  });

  const colorFill = useColorModeValue(
    theme.colors.white,
    colors.secondary[400],
  );

  const handleSend = useCallback(
    async (data: TMessages) => {
      const filterMessage = data.text.replace(/<\/?[^>]+(>|$)/g, '');

      const dataMessage: TMessages = {
        ...data,
        text: filterMessage,
      };

      const idRoomChat = `${currentUser?.uid}${userUid}`;
      const adminId = currentUser?.uid as string;

      await sendMessage(
        dataMessage,
        idRoomChat,
        userUid || '',
        adminId,
        avatarUser,
        currentUser?.avatarURL || '',
        nameUser,
      );

      reset();
    },
    [currentUser, userUid, avatarUser, nameUser, reset],
  );

  const handleOnKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === 'Enter' && !event.shiftKey) {
        event.preventDefault();

        handleSubmit(handleSend)();
        reset();
      }
    },
    [handleSubmit, handleSend, reset],
  );

  useEffect(() => {
    const tooltip = document.querySelector('.ql-tooltip');
    tooltip?.remove();
  });

  return (
    <VStack as="form" onSubmit={handleSubmit(handleSend)}>
      <Flex direction="column" width="full">
        <Controller
          control={control}
          name="text"
          render={({ field: { onChange, ...rest } }) => (
            <Flex direction="row" alignItems="center">
              <Flex width="full" alignItems="center" justify="flex-start">
                <ReactQuill
                  {...rest}
                  onKeyDown={handleOnKeyDown}
                  onChange={onChange}
                  modules={{
                    toolbar: [
                      ['bold', 'italic', 'underline'],
                      ['image', 'code-block'],
                    ],
                  }}
                  style={{
                    width: '100%',
                    backgroundColor: colorFill,
                  }}
                  theme="snow"
                />
              </Flex>
            </Flex>
          )}
        />

        <Text color="red" textAlign="center" py={2} h={10}>
          {root?.message}
        </Text>
        <CustomButton
          w="100px"
          h="40px"
          px={4}
          ml={4}
          py={2.5}
          leftIcon={<SendIconLight />}
          fontSize="md"
          fontWeight="semibold"
          bgColor="primary.600"
          alignSelf="flex-end"
          type="submit"
          isDisabled={!isValid || isSubmitting}
        >
          Send
        </CustomButton>
      </Flex>
    </VStack>
  );
};

export default Quill;
