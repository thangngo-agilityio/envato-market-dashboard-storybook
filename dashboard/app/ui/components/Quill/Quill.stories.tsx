import { StoryObj, Meta } from '@storybook/react';

// Components
import { Quill } from '@/ui/components';

const meta: Meta<typeof Quill> = {
  title: 'Custom Components/Quill',
  tags: ['autodocs'],
  component: Quill,
};

type Story = StoryObj<typeof Quill>;

export const Default: Story = {
  args: {
    userUid: '1',
  },
};

export default meta;
