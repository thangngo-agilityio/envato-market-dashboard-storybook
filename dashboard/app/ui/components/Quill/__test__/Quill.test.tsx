import preloadAll from 'jest-next-dynamic';
import Quill from '..';
import userEvent from '@testing-library/user-event';
import { MOCK_USER_DETAIL } from '@/lib/mocks';
import { IMAGES } from '@/lib/constants';

const sendMessageMock = jest.fn();
const authStoreMock = jest.fn();

jest.mock('@/lib/utils', () => ({
  ...jest.requireActual('@/lib/utils'),
  sendMessage: () => sendMessageMock,
}));

jest.mock('@/lib/stores', () => ({
  ...jest.requireActual('@/lib/stores'),
  authStore: () => authStoreMock,
}));

describe('Quill component', () => {
  beforeAll(async () => {
    await preloadAll();
    authStoreMock.mockReturnValue({
      user: MOCK_USER_DETAIL,
    });
  });
  it('renders correctly', () => {
    const { container } = render(
      <Quill userUid="1" avatarUser={IMAGES.AVATAR.url} nameUser="Loc Vo " />,
    );
    expect(container).toMatchSnapshot();
  });

  it('handle send when enter', async () => {
    const { container } = render(
      <Quill userUid="1" avatarUser={IMAGES.AVATAR.url} nameUser="Loc Vo " />,
    );
    const quill = container.querySelector('.quill');
    if (quill) {
      await userEvent.type(quill, 'hello');
      fireEvent.keyDown(quill, { key: 'Enter', code: 'Enter' });
      waitFor(() => {
        expect(sendMessageMock).toHaveBeenCalled();
      });
    }
  });
});
