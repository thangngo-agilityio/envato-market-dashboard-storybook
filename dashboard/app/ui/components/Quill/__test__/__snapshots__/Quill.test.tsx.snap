// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quill component renders correctly 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="css-1nlrkd1"
    >
      <div
        class="css-u4p24i"
      >
        <div
          class="css-14ethi5"
        >
          <div
            class="quill "
            style="width: 100%; background-color: rgb(255, 255, 255);"
          >
            <div
              class="ql-toolbar ql-snow"
            >
              <span
                class="ql-formats"
              >
                <button
                  class="ql-bold"
                  type="button"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <path
                      class="ql-stroke"
                      d="M5,4H9.5A2.5,2.5,0,0,1,12,6.5v0A2.5,2.5,0,0,1,9.5,9H5A0,0,0,0,1,5,9V4A0,0,0,0,1,5,4Z"
                    />
                     
                    <path
                      class="ql-stroke"
                      d="M5,9h5.5A2.5,2.5,0,0,1,13,11.5v0A2.5,2.5,0,0,1,10.5,14H5a0,0,0,0,1,0,0V9A0,0,0,0,1,5,9Z"
                    />
                     
                  </svg>
                </button>
                <button
                  class="ql-italic"
                  type="button"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <line
                      class="ql-stroke"
                      x1="7"
                      x2="13"
                      y1="4"
                      y2="4"
                    />
                     
                    <line
                      class="ql-stroke"
                      x1="5"
                      x2="11"
                      y1="14"
                      y2="14"
                    />
                     
                    <line
                      class="ql-stroke"
                      x1="8"
                      x2="10"
                      y1="14"
                      y2="4"
                    />
                     
                  </svg>
                </button>
                <button
                  class="ql-underline"
                  type="button"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <path
                      class="ql-stroke"
                      d="M5,3V9a4.012,4.012,0,0,0,4,4H9a4.012,4.012,0,0,0,4-4V3"
                    />
                     
                    <rect
                      class="ql-fill"
                      height="1"
                      rx="0.5"
                      ry="0.5"
                      width="12"
                      x="3"
                      y="15"
                    />
                     
                  </svg>
                </button>
              </span>
              <span
                class="ql-formats"
              >
                <button
                  class="ql-image"
                  type="button"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <rect
                      class="ql-stroke"
                      height="10"
                      width="12"
                      x="3"
                      y="4"
                    />
                     
                    <circle
                      class="ql-fill"
                      cx="6"
                      cy="7"
                      r="1"
                    />
                     
                    <polyline
                      class="ql-even ql-fill"
                      points="5 12 5 11 7 9 8 10 11 7 13 9 13 12 5 12"
                    />
                     
                  </svg>
                </button>
                <button
                  class="ql-code-block"
                  type="button"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <polyline
                      class="ql-even ql-stroke"
                      points="5 7 3 9 5 11"
                    />
                     
                    <polyline
                      class="ql-even ql-stroke"
                      points="13 7 15 9 13 11"
                    />
                     
                    <line
                      class="ql-stroke"
                      x1="10"
                      x2="8"
                      y1="5"
                      y2="13"
                    />
                     
                  </svg>
                </button>
              </span>
            </div>
            <div
              class="ql-container ql-snow"
            >
              <div
                class="ql-editor ql-blank"
                contenteditable="true"
                data-gramm="false"
              >
                <p>
                  <br />
                </p>
              </div>
              <div
                class="ql-clipboard"
                contenteditable="true"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
      <p
        class="chakra-text css-rj1t1b"
      />
      <button
        class="chakra-button css-9xp0iw"
        disabled=""
        type="submit"
      >
        <span
          class="chakra-button__icon css-1wh2kri"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.0586 7.09154L7.92522 3.52487C3.13355 1.12487 1.16689 3.09153 3.56689 7.8832L4.29189 9.3332C4.50022 9.7582 4.50022 10.2499 4.29189 10.6749L3.56689 12.1165C1.16689 16.9082 3.12522 18.8749 7.92522 16.4749L15.0586 12.9082C18.2586 11.3082 18.2586 8.69153 15.0586 7.09154ZM12.3669 10.6249H7.86689C7.52522 10.6249 7.24189 10.3415 7.24189 9.99987C7.24189 9.6582 7.52522 9.37487 7.86689 9.37487H12.3669C12.7086 9.37487 12.9919 9.6582 12.9919 9.99987C12.9919 10.3415 12.7086 10.6249 12.3669 10.6249Z"
              fill="white"
            />
          </svg>
        </span>
        Send
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
